### ISSUE TITLE
guestGetIpAddress() Not returning the Guest IP Address

### ISSUE THREAD

**Issue by dzflack** [josenk/terraform-provider-esxi#101]
**Describe the bug**

TL;DR: terraform-provider-esxi, after having provisioned an `esxi_guest` resource, does not correctly return the the `esxi_guest.ip_address`.

This is as a result of the guestGetIpAddress() function in [guest_functions.go](https://github.com/josenk/terraform-provider-esxi/blob/7ab909a3c0624ea6e949f60e19f2373250b920df/esxi/guest_functions.go#L457).

This function executes the following shell [command](https://github.com/josenk/terraform-provider-esxi/blob/7ab909a3c0624ea6e949f60e19f2373250b920df/esxi/guest_functions.go#L475) (via SSH) on the esxi host, in an attempt to retrieve the newly created esxi_guest's ip address. 

The second and third commands in the pipe chain (i.e.: `grep -A 5 'deviceConfigId = 4000' |tail -1`), do not correctly filter the output of the `vim-cmd` command.

After having executed `terraform apply`, and then SSH'd to the ESXI host, if I manually run  `vim-cmd vmsvc/get.guest 168 2>/dev/null` (the first command in the command chain), the following output is observed:

```
Guest information:

(vim.vm.GuestInfo) {
   toolsStatus = "toolsOk",
   toolsVersionStatus = "guestToolsCurrent",
   toolsVersionStatus2 = "guestToolsCurrent",
   toolsRunningStatus = "guestToolsRunning",
   toolsVersion = "11265",
   toolsInstallType = "guestToolsTypeMSI",
   toolsUpdateStatus = (vim.vm.GuestInfo.ToolsUpdateStatus) null,
   guestId = "windows8Server64Guest",
   guestFamily = "windowsGuest",
   guestFullName = "Microsoft Windows Server 2012 (64-bit)",
   hostName = "vagrant-2012r2",
   ipAddress = "192.168.1.37",
   net = (vim.vm.GuestInfo.NicInfo) [
      (vim.vm.GuestInfo.NicInfo) {
         network = "LAN",
         ipAddress = (string) [
            "fe80::fce4:942:190b:63f",
            "192.168.1.37"
         ],
         macAddress = "00:02:b3:06:b3:90",
         connected = true,
         deviceConfigId = 4000,
         dnsConfig = (vim.net.DnsConfigInfo) {
            dhcp = false,
            hostName = "",
            domainName = "localdomain",
            ipAddress = (string) [
               "192.168.1.1"
            ],
            searchDomain = <unset>
         },
         ipConfig = (vim.net.IpConfigInfo) {
            ipAddress = (vim.net.IpConfigInfo.IpAddress) [
               (vim.net.IpConfigInfo.IpAddress) {
                  ipAddress = "fe80::fce4:942:190b:63f",
                  prefixLength = 64,
                  origin = "linklayer",
                  state = "unknown",
                  lifetime = <unset>
               },
               (vim.net.IpConfigInfo.IpAddress) {
                  ipAddress = "192.168.1.37",
                  prefixLength = 24,
                  origin = "dhcp",
                  state = "preferred",
                  lifetime = <unset>
               }
            ],
            dhcp = (vim.net.DhcpConfigInfo) {
               ipv6 = (vim.net.DhcpConfigInfo.DhcpOptions) {
                  enable = false,
                  config = <unset>
               },
               ipv4 = (vim.net.DhcpConfigInfo.DhcpOptions) {
                  enable = true,
                  config = <unset>
               }
            },
            autoConfigurationEnabled = <unset>
         },
         netBIOSConfig = (vim.net.NetBIOSConfigInfo) null
      }
   ],
   ipStack = (vim.vm.GuestInfo.StackInfo) [
      (vim.vm.GuestInfo.StackInfo) {
         dnsConfig = (vim.net.DnsConfigInfo) {
            dhcp = false,
            hostName = "vagrant-2012r2",
            domainName = "",
            ipAddress = (string) [
               "192.168.1.1"
            ],
            searchDomain = <unset>
         },
         ipRouteConfig = (vim.net.IpRouteConfigInfo) {
            ipRoute = (vim.net.IpRouteConfigInfo.IpRoute) [
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "0.0.0.0",
                  prefixLength = 0,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = "192.168.1.1",
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "192.168.1.0",
                  prefixLength = 24,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "192.168.1.37",
                  prefixLength = 32,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "192.168.1.255",
                  prefixLength = 32,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "224.0.0.0",
                  prefixLength = 4,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "255.255.255.255",
                  prefixLength = 32,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "::",
                  prefixLength = 0,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = "fe80::1:1",
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "fe80::",
                  prefixLength = 64,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "fe80::fce4:942:190b:63f",
                  prefixLength = 128,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               },
               (vim.net.IpRouteConfigInfo.IpRoute) {
                  network = "ff00::",
                  prefixLength = 8,
                  gateway = (vim.net.IpRouteConfigInfo.Gateway) {
                     ipAddress = <unset>,
                     device = "0"
                  }
               }
            ]
         },
         ipStackConfig = <unset>,
         dhcpConfig = (vim.net.DhcpConfigInfo) null
      }
   ],
   disk = (vim.vm.GuestInfo.DiskInfo) [
      (vim.vm.GuestInfo.DiskInfo) {
         diskPath = "C:\",
         capacity = 64422408192,
         freeSpace = 54219526144
      }
   ],
   screen = (vim.vm.GuestInfo.ScreenInfo) {
      width = 1024,
      height = 768
   },
   guestState = "running",
   powerPolicy = (vim.vm.PowerPolicy) null,
   appHeartbeatStatus = "appStatusGray",
   guestKernelCrashed = false,
   appState = "none",
   guestOperationsReady = true,
   interactiveGuestOperationsReady = false,
   guestStateChangeSupported = true,
   generationInfo = <unset>,
   customizationInfo = (vim.vm.GuestInfo.CustomizationInfo) null
}
```

If i then pipe that outpt into  `grep -A 5 'deviceConfigId = 4000' |tail -1` (which is what the function currently does),it removes the  value that we need (`ipAddress = "192.168.1.37"`), and outputs the following: 

```
deviceConfigId = 4000,
dnsConfig = (vim.net.DnsConfigInfo) {
dhcp = false,
hostName = "",
domainName = "localdomain",
ipAddress = (string) [
```

I was able to get it working by replacing those two commands with `grep 'ipAddress' |head -1`:

```go
remote_cmd = fmt.Sprintf("vim-cmd vmsvc/get.guest %s 2>/dev/null |grep 'ipAddress' |head -1|grep -oE '((1?[0-9][0-9]?|2[0-4][0-9]|25[0-5]).){3}(1?[0-9][0-9]?|2[0-4][0-9]|25[0-5])'", vmid)
```

The above is a quick fix, and I'm sure there is a better way of going about it.

**To Reproduce**
1. terraform plan
2. terraform apply
2. observe the terraform output value for "esxi_guest.dc1.ip_address" is a blank string. i.e.:

```
esxi_guest.dc1: Creation complete after 7m54s [id=168]
Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
Outputs:
dc1 = [
  "",
]
```

**Expected behavior**
I expected the terraform output value for "esxi_guest.dc1.ip_address" to contain the ip address of the new esxi guest. i.e.:

```
esxi_guest.dc1: Creation complete after 7m54s [id=168]
Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
Outputs:
dc1 = [
  "192.168.1.37",
]
```

**Terraform files**

main.tf:

```
terraform {
  required_version = ">= 0.12"
}

provider "esxi" {
  esxi_hostname      = var.esxi_hostname
  esxi_hostport      = var.esxi_hostport
  esxi_username      = var.esxi_username
  esxi_password      = var.esxi_password
}

# dc1

resource "esxi_guest" "dc1" {
  guest_name         = var.dcname
  disk_store         = var.disk_store

  numvcpus = "2"
  memsize = "2048"
  power   = "on"

  clone_from_vm = var.template_image

  network_interfaces {
    virtual_network = var.vm_network
    mac_address = var.dcmac
  }

  guest_startup_timeout = 60
  guest_shutdown_timeout = 30

}
```

outputs.tf:

```
output "dc1" {
  value = [esxi_guest.dc1.ip_address]
}
```

variables.tf:

```
variable "template_image" {
  default = "windows_2012r2-base"
}
variable "esxi_hostname"  { default = "<snipped>" }
variable "esxi_hostport"  { default = "22" }
variable "esxi_username"  { default = "<snipped>" }
variable "esxi_password"  { default = "<snipped>" }
variable "disk_store"     { default = "Local" }
variable "vm_network"     { default = "LAN" }
variable "dcip"  { default = "192.168.1.37" }
variable "dcmac" { default = "00:02:B3:06:B3:90" }
variable "dcname"  { default = "dc1" }
```

**Desktop**
 - OS: `Linux ubuntu1810-packer 4.18.0-25-generic 26-Ubuntu SMP Mon Jun 24 09:32:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
 - Terraform Version: `Terraform v0.12.18`
 - Ovftool Version: `VMware ovftool 4.3.0 (build-13981069)`
 - This Plugin Version: `provider.esxi v1.6.2`

**Comment by josenk** [comment]
You,ll need to provide much more info...   I'll need to recreate the problem before accepting this.

**Comment by dzflack** [comment]
Apologies @josenk I have filled out the template correctly this time.

**Comment by josenk** [comment]
Thank you very much for the `vim-cmd vmsvc/get.guest` output.  It certainly helped clarify things.  Basically, I need to get the IPv4 address from the block containing `deviceConfigId = 4000`. (That block is the first configured virtual_network.)  What makes this tricky to parse is esxi regex is greedy and limited, so I need to get creative to capture the IPv4 information.

I found that your output does not follow all the other examples I've looked at.  There are certain things that are out of order.  

  * In my examples, I always searched for `deviceConfigId = 4000`, found the `ipConfig = (vim.net.IpConfigInfo)` block, then the `ipAddress = (vim.net.IpConfigInfo.IpAddress)` block containing the IPv4 first.
  * In your example, following `deviceConfigId = 4000` is the `dnsConfig = (vim.net.DnsConfigInfo)` block.  Also, the first IP address in the `ipConfig = (vim.net.IpConfigInfo)` is IPv6!   I'm not sure why your block is a bit different?


With a bit of creativity, I was able to use a different technique to get the (first) IPv4 address from the `deviceConfigId = 4000` block.  I tested it with my system your example and it seems to work.  Before I implement it in the next release can you please test the following on as many esxi hosts as possible to verify it always returns the correct IPv4 result.

```
for i in `vim-cmd vmsvc/getallvms |tail -n +2 |awk '{print $1}'`
do
  name=`vim-cmd vmsvc/get.summary $i |grep "name = "`
  echo -e "$i  $name: \c"
  vim-cmd vmsvc/get.guest $i | sed '1!G;h;$!d' |awk '/deviceConfigId = 4000/,/ipAddress/'|grep -oE '((1?[0-9][0-9]?|2[0-4][0-9]|25[0-5])\.){3}(1?[0-9][0-9]?|2[0-4][0-9]|25[0-5])' || echo
done
```

**Comment by dzflack** [comment]
Not sure why the output I provided is different to the other examples you've seen before, could it be due to the image having been built using Packer?

Running the above script does correctly return all the IP addresses of the guests, however, I did have a question as to why you needed to specifically obtain the IP address from the first configured virtual network? (i.e. the `deviceConfigId = 4000` block). Is this method something you have identified through trial and error? Rummaging through VMWare doco I came across [this](https://kb.vmware.com/s/article/2012964) page, which suggests obtaining the Guest IP via the following command: `vim-cmd vmsvc/get.guest `*`vmid`*` |grep -m 1 "ipAddress = \""`.

**Comment by josenk** [comment]
I'm not 100% sure why your output is a different???   Maybe it's packer.   I guess it doesn't really matter why.   If it's happening out there and the output is valid, then I need to deal with it...

That VMware doc just points to the "global" IPaddress.   That IPaddress is what I use as the "alternate method", when I'm unable to get an IP from the primary nic.
>https://github.com/josenk/terraform-provider-esxi/blob/master/esxi/guest_functions.go#L496

The reason why I choose block 4000 (the first network interface) is because I found that global IP address is sometimes inconsistent when you have multiple network interfaces.   I have a few boxes that pull the IP from the second nic (block 4001).   I prefer consistency and the ability to choose my primary IP.

Thanks for testing the code...

**Comment by dzflack** [comment]
Ahhh yep makes sense. No worries, let me know if there is anything else I can help with!

**Comment by josenk** [comment]
Should be fixed in latest release.
> https://github.com/josenk/terraform-provider-esxi/releases/tag/v1.6.4

**Comment by dzflack** [comment]
Awesome, thanks for getting the fix out so promptly! :)
